topiary/predictor.py

```diff
@@ -159,26 +228,123 @@ def epitopes_from_mutation_effects(
+            ('effect_type', []),
+            ('contains_mutant_residues', []),
+            ('mutation_start_in_peptide', []),
+            ('mutation_end_in_peptide', []),
+        ])
+        if gene_expression_dict is not None:
+            extra_columns["gene_expression"] = []
+        if transcript_expression_dict is not None:
+            extra_columns["transcript_expression"] = []
+
+        for _, row in df.iterrows():
+            effect = variant_string_to_effect_dict[row.variant]
+            mutation_start_in_protein = effect.aa_mutation_start_offset
+            mutation_end_in_protein = effect.aa_mutation_end_offset
+            peptide_length = len(row.peptide)
+            is_mutant = contains_mutant_residues(
+                peptide_start_in_protein=row.peptide_offset,
+                peptide_length=peptide_length,
+                mutation_start_in_protein=mutation_start_in_protein,
+                mutation_end_in_protein=mutation_end_in_protein)
+            if is_mutant:
+                mutation_start_in_peptide, mutation_end_in_peptide = peptide_mutation_interval(
+                    peptide_start_in_protein=row.peptide_offset,
+                    peptide_length=peptide_length,
+                    mutation_start_in_protein=mutation_start_in_protein,
+                    mutation_end_in_protein=mutation_end_in_protein)
+            else:
+                mutation_start_in_peptide = mutation_end_in_peptide = None
+
+            extra_columns["gene"].append(effect.gene_name)
+            gene_id = effect.gene_id
+            extra_columns["gene_id"].append(gene_id)
+            if gene_expression_dict is not None:
+                extra_columns["gene_expression"].append(
+                    gene_expression_dict.get(gene_id, 0.0))
+
+            transcript_id = effect.transcript_id
+            extra_columns["transcript_id"].append(transcript_id)
+            extra_columns["transcript_name"].append(effect.transcript_name)
+            if transcript_expression_dict is not None:
+                extra_columns["transcript_expression"].append(
+                    transcript_expression_dict.get(transcript_id, 0.0))
+
+            extra_columns["effect"].append(effect.short_description)
+            extra_columns["effect_type"].append(effect.__class__.__name__)
+
+            extra_columns["contains_mutant_residues"].append(is_mutant)
+            extra_columns["mutation_start_in_peptide"].append(mutation_start_in_peptide)
+            extra_columns["mutation_end_in_peptide"].append(mutation_end_in_peptide)
+
+        for col, values in extra_columns.items():
+            df[col] = values
+
+        # TODO: add extra boolean field
+        #   novel = is_mutant | not_in_reference
+        # Requires keeping a quick lookup structure for all peptides in
+        # the reference proteome
+        if self.only_novel_epitopes:
+            df = df[df.contains_mutant_residues]
+
+        return df
+
+    def predict_variants(
```

Comment:
Weird name, sounds like you're predicting variants. Maybe `predict_from_variants`?